src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -291,18 +293,35 @@ private Map<Integer, Float> doExactSearch(final LeafReaderContext leafReaderCont
         try {
             final BinaryDocValues values = DocValues.getBinary(leafReaderContext.reader(), fieldInfo.name);
             final SpaceType spaceType = SpaceType.getSpace(fieldInfo.getAttribute(SPACE_TYPE));
-
+            final HitQueue queue = new HitQueue(this.knnQuery.getK(), true);
+            ScoreDoc topDoc = queue.top();
             final Map<Integer, Float> docToScore = new HashMap<>();
-            for (int j : filterIdsArray) {
-                int docId = values.advance(j);
-                BytesRef value = values.binaryValue();
-                ByteArrayInputStream byteStream = new ByteArrayInputStream(value.bytes, value.offset, value.length);
+            for (int filterId : filterIdsArray) {
+                int docId = values.advance(filterId);
+                final BytesRef value = values.binaryValue();
+                final ByteArrayInputStream byteStream = new ByteArrayInputStream(value.bytes, value.offset,
+                        value.length);
                 final KNNVectorSerializer vectorSerializer = KNNVectorSerializerFactory.getSerializerByStreamContent(byteStream);
                 final float[] vector = vectorSerializer.byteToFloatArray(byteStream);
-                // making min score as high score as this is closest to the vector
+                // Calculates a similarity score between the two vectors with a specified function. Higher similarity
+                // scores correspond to closer vectors.
                 float score = spaceType.getVectorSimilarityFunction().compare(queryVector, vector);
-                docToScore.put(docId, score);
+                if(score > topDoc.score) {
+                    topDoc.score = score;
+                    topDoc.doc = docId;
+                    topDoc = queue.updateTop();
+                }
+            }
+            // If scores are negative we will remove them
+            while (queue.size() > 0 && queue.top().score < 0) {
```

Comment:
If queue.top() score is negative, doesn't that mean every score in the queue is negative? Then, we can simply empties it instead of looping through it?

Reply:
Given that heap is init with all values having score of -INF. So in case filterIds < k, some ids in the heap can have -INF value. Hence we need to remove them.